MailAlert.php

```diff
@@ -304,13 +305,17 @@ public static function getProductAttributeCombination($id_product_attribute, $id
 
     /*
      * Get customers waiting for alert on the specified product/product attribute
+     * in shop `$id_shop` and if the shop group shares the stock in all shops of the shop group
      */
-    public static function getCustomers($id_product, $id_product_attribute)
+    public static function getCustomers($id_product, $id_product_attribute, $id_shop)
     {
         $sql = '
-			SELECT id_customer, customer_email, id_shop, id_lang
-			FROM `' . _DB_PREFIX_ . self::$definition['table'] . '`
-			WHERE `id_product` = ' . (int) $id_product . ' AND `id_product_attribute` = ' . (int) $id_product_attribute;
+			SELECT mc.id_customer, mc.customer_email, mc.id_shop, mc.id_lang
+			FROM `' . _DB_PREFIX_ . self::$definition['table'] . '` mc
+			INNER JOIN `' . _DB_PREFIX_ . 'shop` s on s.id_shop = mc.id_shop
```

Comment:
I'm not sure about these inner joins, is there a particularly reason to use them rather than left joins? 😊

Reply:
I'm choosing the notifications that are either of the context shop OR of a shop in the same shop group under the condition that the shop group shares the stock. These two choices do not need the same joins. 

For the simple case `the notifications that are of the context shop`, this would be sufficient:
```
    FROM `' . _DB_PREFIX_ . self::$definition['table'] . '` mc
-   WHERE mc.`id_product` = ' . (int) $id_product . ' AND mc.`id_product_attribute` = ' . (int) $id_product_attribute';
+   WHERE mc.id_shop = ' . (int) $id_shop . ' AND mc.`id_product` = ' . (int) $id_product . ' AND mc.`id_product_attribute` = ' . (int) $id_product_attribute';
```

But when we take into account the case `the notifications a shop in the same shop group that the shop group shares the stock`, things become more complicated:
```
    SELECT mc.id_customer, mc.customer_email, mc.id_shop, mc.id_lang
    FROM `' . _DB_PREFIX_ . self::$definition['table'] . '` mc
    INNER JOIN `' . _DB_PREFIX_ . 'shop` s on s.id_shop = mc.id_shop
    INNER JOIN `' . _DB_PREFIX_ . 'shop_group` sg on s.id_shop_group = sg.id_shop_group and (s.id_shop = ' . (int) $id_shop . ' or sg.share_stock = 1)
    INNER JOIN `' . _DB_PREFIX_ . 'shop` s2 on s2.id_shop = mc.id_shop and s2.id_shop = ' . (int) $id_shop . '
    WHERE mc.`id_product` = ' . (int) $id_product . ' AND mc.`id_product_attribute` = ' . (int) $id_product_attribute
```
If the stock is not shared, we can simplify to (removing one join and removing the `share_stock` condition)
```
    INNER JOIN `' . _DB_PREFIX_ . 'shop` s on s.id_shop = mc.id_shop
    INNER JOIN `' . _DB_PREFIX_ . 'shop_group` sg on s.id_shop_group = sg.id_shop_group and s.id_shop = ' . (int) 
    WHERE (...)
```
which could be further be simplified to get the __sufficient query__ above.

If the stock is shared, the query must be:
```
    SELECT mc.id_customer, mc.customer_email, mc.id_shop, mc.id_lang
    FROM `' . _DB_PREFIX_ . self::$definition['table'] . '` mc
    INNER JOIN `' . _DB_PREFIX_ . 'shop` s on s.id_shop = mc.id_shop
    INNER JOIN `' . _DB_PREFIX_ . 'shop_group` sg on s.id_shop_group = sg.id_shop_group and sg.share_stock
    INNER JOIN `' . _DB_PREFIX_ . 'shop` s2 on s2.id_shop = mc.id_shop and s2.id_shop = ' . (int) $id_shop . '
    WHERE mc.`id_product` = ' . (int) $id_product . ' AND mc.`id_product_attribute` = ' . (int) $id_product_attribute
```
To take into account both cases, the new query in the PR is needed. It has to many joins if the shop group does not share the stock, but all queries are needed if the shop group shares the stock.

I preferred to write only one query instead of 1) testing for shares stock, and then 2) run either one of the two queries for the two cases.

Reply:
Thanks a lot for your explanations! 😊 

Have you tried if it all works correctly? It could be very tough to test it :/

Reply:
@marsaldev
I have it in production, and I tested it with the full procedure at https://github.com/PrestaShop/PrestaShop/issues/30187 (I don't have a shop group that shares stock in production).